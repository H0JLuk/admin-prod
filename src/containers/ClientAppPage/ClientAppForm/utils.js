import { notification } from 'antd';
import { addSettings, updateSettingsList } from '../../../api/services/settingsService';
import { designKeysForCheck, SETTINGS_TYPES } from './ClientAppFormConstants';

export const showNotify = (message, isError) => {
    const config = {
        duration: 5,
        placement: 'bottomRight',
    };

    if (isError) {
        notification.error({
            ...config,
            message,
        });
    } else {
        notification.success({
            ...config,
            message,
        });
    }

};

export async function createOrUpdateKey (changedParams) {
    const updateSettings = [];
    const addSettingsArr = [];

    changedParams.forEach(({ type, ...params }) => {
        type === SETTINGS_TYPES.EDIT
            ? updateSettings.push(params)
            : addSettingsArr.push(params);
    });

    updateSettings.length && (await updateSettingsList(updateSettings));
    addSettingsArr.length && (await addSettings(addSettingsArr));
}

export const checkExistDesignSettings = (settings) => {
    return designKeysForCheck.some(key => !Object.prototype.hasOwnProperty.call(settings, key));
};

export const getImage = (imageName) => {
    return `${process.env.PUBLIC_URL}/images/clientAppDesign/${imageName}.png`;
};


