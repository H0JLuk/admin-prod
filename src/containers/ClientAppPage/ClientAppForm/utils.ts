import { customNotifications } from '@utils/notifications';
import { ArgsProps } from 'antd/lib/notification';
import { addSettings, updateSettingsList } from '@apiServices/settingsService';
import { SETTINGS_TYPES } from './ClientAppFormConstants';
import { SettingDto } from '@types';
import ROLES from '@constants/roles';

export interface IChangedParam extends Pick<SettingDto, 'clientAppCode' | 'key' | 'value' | 'userRole'> {
    type: SETTINGS_TYPES;
}

export const showNotify = (message: React.ReactNode, isError?: boolean) => {
    const config: ArgsProps = {
        duration: 5,
        message,
    };

    if (isError) {
        customNotifications.error(config);
    } else {
        customNotifications.success(config);
    }

};

type UpdateAndAddSettings = {
    updateSettings: SettingDto[];
    addSettingsArr: SettingDto[];
};

export async function createOrUpdateKey(changedParams: IChangedParam[]) {
    const { updateSettings, addSettingsArr } = changedParams.reduce<UpdateAndAddSettings>((result, { type, ...params }) => {
        const key = type === SETTINGS_TYPES.EDIT ? 'updateSettings' : 'addSettingsArr';

        if (params.key === 'referralTokenLifetime') {
            delete params.clientAppCode;
            params.key = 'token_lifetime';
            params.userRole = ROLES.REFERAL_LINK;
        }

        return {
            ...result,
            [key]: [...result[key], params],
        };
    }, { updateSettings: [], addSettingsArr: [] });

    updateSettings.length && (await updateSettingsList(updateSettings));
    addSettingsArr.length && (await addSettings(addSettingsArr));
}
