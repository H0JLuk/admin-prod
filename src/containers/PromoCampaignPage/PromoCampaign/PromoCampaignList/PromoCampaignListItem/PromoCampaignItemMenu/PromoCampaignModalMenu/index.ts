import UploadPromoCodesModal from './UploadPromoCodesModal';
import PromoCodeStatisticModal from './PromoCodeStatisticModal';

export {
    UploadPromoCodesModal,
    PromoCodeStatisticModal,
};

export default UploadPromoCodesModal;
