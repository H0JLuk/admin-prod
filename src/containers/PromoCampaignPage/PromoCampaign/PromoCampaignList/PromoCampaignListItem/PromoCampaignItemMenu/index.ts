import PromoCampaignItemMenu from './PromoCampaignItemMenu';

export default PromoCampaignItemMenu;
