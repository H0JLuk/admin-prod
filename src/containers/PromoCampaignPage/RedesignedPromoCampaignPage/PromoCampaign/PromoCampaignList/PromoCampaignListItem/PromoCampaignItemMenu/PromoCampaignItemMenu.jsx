import React, { useCallback, useState } from 'react';
import { generatePath, Link } from 'react-router-dom';
import { Menu, Dropdown } from 'antd';
import { EllipsisOutlined } from '@ant-design/icons';
import { deletePromoCampaign } from '../../../../../../../api/services/promoCampaignService';
import { PROMO_CAMPAIGN_PAGES } from '../../../../../../../constants/route';

import styles from './PromoCampaignItemMenu.module.css';

const PromoCampaignItemMenu = ({ onDeleteItem, promoCampaign, matchUrl }) => {
    const [dropdownVisible, setDropdownVisible] = useState(false);

    const onDeletePromoCampaign = useCallback(async (e) => {
        e.stopPropagation();
        try {
            await deletePromoCampaign(promoCampaign.id);
            onDeleteItem();
        } catch (e) {
            // TODO: add error handler
            console.error(e);
        }

        setDropdownVisible(false);
    }, [promoCampaign.id, onDeleteItem]);

    const onMenuClick = useCallback((e) => e.stopPropagation(), []);

    return (
        <Dropdown
            className={ styles.dropdown }
            overlayClassName={ styles.dropdownMenu }
            overlay={
                <DropdownMenu
                    promoCampaign={ promoCampaign }
                    onDelete={ onDeletePromoCampaign }
                    onClickItem={ onMenuClick }
                    matchUrl={ matchUrl }
                />
            }
            trigger={ ['click'] }
            visible={ dropdownVisible }
            onVisibleChange={ setDropdownVisible }
        >
            <div className={ styles.menuButton } onClick={ onMenuClick }>
                <EllipsisOutlined />
            </div>
        </Dropdown>
    );
};


export default PromoCampaignItemMenu;


const MENU = {
    LOAD_PROMO_CODE: 'Загрузить промо-коды',
    STATISTICS: 'Статистика использования промо-кодов',
    VISIBILITY_SETTINGS: 'Настроить видимость промо-кампании',
    EDIT_PROMO_CAMPAIGN: 'Редактировать промо-кампанию',
    DELETE: 'Удалить',
};

const DropdownMenu = ({ onDelete, onClickItem, promoCampaign, matchUrl }) => (
    <Menu>
        <Menu.Item key="0">
            <Link
                onClick={ onClickItem }
                to={ generatePath(
                    `${ matchUrl }${PROMO_CAMPAIGN_PAGES.LOAD_PROMO_CODES}`,
                    { promoCampaignId: promoCampaign.id }
                ) }
            >
                { MENU.LOAD_PROMO_CODE }
            </Link>
        </Menu.Item>
        <Menu.Item key="1">
            <Link
                onClick={ onClickItem }
                to={ generatePath(
                    `${ matchUrl }${PROMO_CAMPAIGN_PAGES.PROMO_CODES_STATISTIC}`,
                    { promoCampaignId: promoCampaign.id }
                ) }
            >
                { MENU.STATISTICS }
            </Link>
        </Menu.Item>
        <Menu.Item key="2">
            <Link
                onClick={ onClickItem }
                to={ generatePath(
                    `${ matchUrl }${PROMO_CAMPAIGN_PAGES.VISIBILITY_SETTINGS}`,
                    { promoCampaignId: promoCampaign.id }
                ) }
            >
                { MENU.VISIBILITY_SETTINGS }
            </Link>
        </Menu.Item>
        <Menu.Item key="3">
            <Link
                onClick={ onClickItem }
                to={ {
                    pathname: generatePath(
                        `${ matchUrl }${PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_EDIT}`,
                        { promoCampaignId: promoCampaign.id }
                    ),
                    state: { promoCampaign },
                } }
            >
                { MENU.EDIT_PROMO_CAMPAIGN }
            </Link>
        </Menu.Item>
        <Menu.Item key="4">
            <div className={ styles.menuItem } onClick={ onDelete }>
                { MENU.DELETE }
            </div>
        </Menu.Item>
    </Menu>
);
