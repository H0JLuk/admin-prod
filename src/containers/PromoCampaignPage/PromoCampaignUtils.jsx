import React from 'react';
import { deletePromoCampaign } from '../../api/services/promoCampaignService';
import { customNotifications } from '../../utils/notifications';


export const getDeleteTitleConfirm = (promoCampaignName) => (
    <span>
        Вы уверены, что хотите удалить промо-кампанию <b>{ promoCampaignName }</b>?
    </span>
);

export const getSuccessDeleteMessage = (promoCampaignName) => (
    <span>
        Промо-кампания <b>{ promoCampaignName }</b> успешно удалена
    </span>
);

export async function onConfirmDeletePromoCampaign(id, name) {
    try {
        await deletePromoCampaign(id);
        customNotifications.success({
            duration: 10,
            message: getSuccessDeleteMessage(name),
        });
    } catch ({ message }) {
        customNotifications.error({
            message,
        });
        throw new Error(message);
    }
}
