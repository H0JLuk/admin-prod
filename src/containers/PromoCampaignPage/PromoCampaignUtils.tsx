import React from 'react';
import { deletePromoCampaign } from '../../api/services/promoCampaignService';
import { customNotifications } from '../../utils/notifications';

export const getDeleteTitleConfirm = (promoCampaignName: string) => (
    <span>
        Вы уверены, что хотите удалить промо-кампанию <b>{promoCampaignName}</b>?
    </span>
);

export const getSuccessDeleteMessage = (promoCampaignName: string) => (
    <span>
        Промо-кампания <b>{promoCampaignName}</b> успешно удалена
    </span>
);

export async function onConfirmDeletePromoCampaign(id: number, name: string) {
    try {
        await deletePromoCampaign(id);
        customNotifications.success({
            duration: 10,
            message: getSuccessDeleteMessage(name),
        });
    } catch ({ message }) {
        customNotifications.error({
            message,
        });
        throw new Error(message);
    }
}
