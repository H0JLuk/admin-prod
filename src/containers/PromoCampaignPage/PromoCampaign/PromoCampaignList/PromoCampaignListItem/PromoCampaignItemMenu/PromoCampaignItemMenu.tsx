import React, { useCallback, useState } from 'react';
import { generatePath, Link } from 'react-router-dom';
import { Menu, Dropdown } from 'antd';
import { EllipsisOutlined } from '@ant-design/icons';
import PromoCodeStatisticModal from './PromoCampaignModalMenu/PromoCodeStatisticModal';
import { errorNotice } from '@components/toast/Notice';
import UploadPromoCodesModal from './PromoCampaignModalMenu';
import PromoCampaignCopyModal from '../../../PromoCampaignCopyModal';
import { uploadPromoCodes } from '@apiServices/promoCampaignService';
import { PROMO_CAMPAIGN_PAGES } from '@constants/route';
import { confirmModal } from '@utils/utils';
import { getDeleteTitleConfirm, onConfirmDeletePromoCampaign } from '../../../../PromoCampaignUtils';
import { PromoCampaignDto } from '@types';
import { customNotifications } from '@utils/notifications';
import { BUTTON_TEXT } from '@constants/common';

import styles from './PromoCampaignItemMenu.module.css';

const ON_PROMO_CODES_LOADED = 'Промокоды успешно загружены';
const NONE_PROMO_CODES = 'NONE';

const showSuccessNotification = (message: React.ReactNode) => {
    customNotifications.success({ message });
};

type PromoCampaignItemMenuProps = {
    onDeleteItem: () => void;
    promoCampaign: PromoCampaignDto;
    matchUrl: string;
};

interface DropdownMenuProps extends Omit<PromoCampaignItemMenuProps, 'onDeleteItem'> {
    onDelete: () => void;
    onShowStatistic: () => void;
    onPromoCodeUpload: () => void;
    closeDropdown: () => void;
}

const PromoCampaignItemMenu: React.FC<PromoCampaignItemMenuProps> = ({ onDeleteItem, promoCampaign, matchUrl }) => {
    const [dropdownVisible, setDropdownVisible] = useState(false);
    const [showModalStatistic, setShowModalStatistic] = useState(false);
    const [showPromoCodesModal, setShowPromoCodesModal] = useState(false);

    const onConfirmDelete = useCallback(async () => {
        try {
            await onConfirmDeletePromoCampaign(promoCampaign.id, promoCampaign.name);
            onDeleteItem();
        } catch (err) {
            console.warn(err);
        }
    }, [promoCampaign.id, promoCampaign.name, onDeleteItem]);

    const closeDropdown = useCallback(() => setDropdownVisible(false), []);

    const onDeleteClick = useCallback(() => {
        confirmModal({
            onOk: onConfirmDelete,
            title: getDeleteTitleConfirm(promoCampaign.name),
        });
    }, [promoCampaign.name, onConfirmDelete]);

    const onShowStatistic = useCallback(() => {
        setShowModalStatistic(true);
    }, []);

    const onHideStatistic = useCallback(() => {
        setShowModalStatistic(false);
    }, []);

    const onPromoCodeUpload = useCallback(() => {
        setShowPromoCodesModal(true);
    }, []);

    const closeUploadPromoCodesModal = useCallback(() => {
        setShowPromoCodesModal(false);
    }, []);

    const promoCodeUpload = useCallback(async (data) => {
        try {
            await uploadPromoCodes(promoCampaign.id, data);
            closeUploadPromoCodesModal();
            showSuccessNotification(ON_PROMO_CODES_LOADED);
        } catch (error) {
            errorNotice(error.message);
        }
    }, [closeUploadPromoCodesModal, promoCampaign.id]);

    return (
        <>
            <Dropdown
                className={styles.dropdown}
                overlayClassName={styles.dropdownMenu}
                overlay={
                    <DropdownMenu
                        promoCampaign={promoCampaign}
                        onDelete={onDeleteClick}
                        onShowStatistic={onShowStatistic}
                        onPromoCodeUpload={onPromoCodeUpload}
                        matchUrl={matchUrl}
                        closeDropdown={closeDropdown}
                    />
                }
                trigger={['click']}
                visible={dropdownVisible}
                onVisibleChange={setDropdownVisible}
            >
                <div className={styles.menuButton}>
                    <EllipsisOutlined />
                </div>
            </Dropdown>
            <PromoCodeStatisticModal
                open={showModalStatistic}
                onClose={onHideStatistic}
                {...promoCampaign}
            />
            <UploadPromoCodesModal
                open={showPromoCodesModal}
                onClose={closeUploadPromoCodesModal}
                onSave={promoCodeUpload}
            />
        </>
    );
};


export default PromoCampaignItemMenu;


const MENU = {
    LOAD_PROMO_CODE: 'Загрузить промо-коды',
    STATISTICS: 'Статистика использования промо-кодов',
    VISIBILITY_SETTINGS: 'Настроить видимость промо-кампании',
    EDIT_PROMO_CAMPAIGN: 'Редактировать промо-кампанию',
    COPY_PROMO: 'Копировать промо-кампанию',
    DELETE: BUTTON_TEXT.DELETE,
};

const DropdownMenu: React.FC<DropdownMenuProps> = ({
    onDelete,
    promoCampaign,
    matchUrl,
    onShowStatistic,
    onPromoCodeUpload,
    closeDropdown,
}) => (
    <Menu onClick={closeDropdown}>
        {promoCampaign.promoCodeType !== NONE_PROMO_CODES && (
            <>
                <Menu.Item key="0">
                    <div className={styles.menuItem} onClick={onPromoCodeUpload}>
                        {MENU.LOAD_PROMO_CODE}
                    </div>
                </Menu.Item>
                <Menu.Item key="1">
                    <div className={styles.menuItem} onClick={onShowStatistic}>
                        {MENU.STATISTICS}
                    </div>
                </Menu.Item>
            </>
        )}
        <Menu.Item key="2">
            <Link
                to={generatePath(
                    `${ matchUrl }${PROMO_CAMPAIGN_PAGES.VISIBILITY_SETTINGS}`,
                    { promoCampaignId: promoCampaign.id },
                )}
            >
                {MENU.VISIBILITY_SETTINGS}
            </Link>
        </Menu.Item>
        <Menu.Item key="3">
            <Link
                to={{
                    pathname: generatePath(
                        `${ matchUrl }${PROMO_CAMPAIGN_PAGES.PROMO_CAMPAIGN_EDIT}`,
                        { promoCampaignId: promoCampaign.id },
                    ),
                    state: { promoCampaign },
                }}
            >
                {MENU.EDIT_PROMO_CAMPAIGN}
            </Link>
        </Menu.Item>
        <Menu.Item key="copyPromoCampaign">
            <PromoCampaignCopyModal promoCampaignData={promoCampaign}>
                <div className={styles.menuItem}>
                    {MENU.COPY_PROMO}
                </div>
            </PromoCampaignCopyModal>
        </Menu.Item>
        <Menu.Item key="4">
            <div className={styles.menuItem} onClick={onDelete}>
                {MENU.DELETE}
            </div>
        </Menu.Item>
    </Menu>
);
