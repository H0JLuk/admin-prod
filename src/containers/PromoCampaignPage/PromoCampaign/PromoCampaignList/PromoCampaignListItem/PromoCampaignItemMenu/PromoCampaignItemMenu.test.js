import React from 'react';
import { shallow } from 'enzyme';
import { message } from 'antd';
import PromoCampaignItemMenu from './PromoCampaignItemMenu';

import * as utils from '../../../../../../utils/utils';
import * as promoCampaignService from '../../../../../../api/services/promoCampaignService';
import * as promoCampaignUtils from '../../../../PromoCampaignUtils';

promoCampaignService.uploadPromoCodes = jest.fn();
utils.confirmModal = jest.fn();
promoCampaignUtils.onConfirmDeletePromoCampaign = jest.fn();

describe('<PromoCampaignItemMenu /> test', () => {
    const initialProps = {
        onDeleteItem: jest.fn(),
        promoCampaign: {
            id: 1,
            name: 'promoCampaign',
            promoCodeType: 'PERSONAL',
        },
        matchUrl: 'url',
    };

    const container = shallow(<PromoCampaignItemMenu { ...initialProps } />);

    it('should match snapshot', () => {
        expect(container.html()).toMatchSnapshot();
    });

    it('confirmModal should be called', () => {
        const dropDownContainer = shallow(container.find('Dropdown').prop('overlay'));
        dropDownContainer.find('div').at(3).simulate('click');
        expect(utils.confirmModal).toBeCalledTimes(1);
    });

    it('dropdown should be displayed', () => {
        container.find('Dropdown').simulate('visibleChange', true);
        expect(container.find('Dropdown').prop('visible')).toBe(true);
    });

    it('dropdown should not be displayed', () => {
        const dropDownContainer = shallow(container.find('Dropdown').prop('overlay'));
        dropDownContainer.find('Menu').simulate('click');
        expect(container.find('Dropdown').prop('visible')).toBe(false);
    });

    it('modal statistic should be displayed', () => {
        container.find('Dropdown').prop('overlay').props.onShowStatistic();
        expect(container.find('PromoCodeStatisticModal').prop('open')).toBe(true);
    });

    it('modal statistic should not be displayed', () => {
        container.find('PromoCodeStatisticModal').prop('onClose')();
        expect(container.find('PromoCodeStatisticModal').prop('open')).toBe(false);
    });

    it('promo codes modal should be displayed', () => {
        container.find('Dropdown').prop('overlay').props.onPromoCodeUpload();
        expect(container.find('UploadPromoCodesModal').prop('open')).toBe(true);
    });

    it('promo codes modal should n`t be displayed', () => {
        container.find('UploadPromoCodesModal').prop('onClose')();
        expect(container.find('UploadPromoCodesModal').prop('open')).toBe(false);
    });

    it('uploadPromoCodes should be called is in a state fulfilled', () => {
        promoCampaignService.uploadPromoCodes.mockResolvedValue();
        container.find('UploadPromoCodesModal').prop('onSave')();
        expect(promoCampaignService.uploadPromoCodes).toHaveBeenCalledTimes(1);
    });

    it('should call confirmModal on deleting', async () => {
        const { promoCampaign } = initialProps;
        container.find('Dropdown').prop('overlay').props.onDelete();

        expect(utils.confirmModal).toHaveBeenCalledTimes(1);

        await utils.confirmModal.mock.calls[0][0].onOk();

        expect(promoCampaignUtils.onConfirmDeletePromoCampaign).toHaveBeenCalledTimes(1);
        expect(promoCampaignUtils.onConfirmDeletePromoCampaign).toHaveBeenCalledWith(
            promoCampaign.id,
            promoCampaign.name
        );
        expect(initialProps.onDeleteItem).toHaveBeenCalledTimes(1);
    });

    it('console.warn should be called when onConfirmDeletePromoCampaign rejected', async () => {
        const spy = jest.spyOn(console, 'warn').mockImplementation(() => '');
        promoCampaignUtils.onConfirmDeletePromoCampaign.mockImplementation(() => Promise.reject('test error'));
        container.find('Dropdown').prop('overlay').props.onDelete();
        await utils.confirmModal.mock.calls[0][0].onOk();
        expect(spy).toHaveBeenCalledTimes(1);
        expect(spy).toHaveBeenCalledWith('test error');
    });

    it('errorNotice should be called when promoCodeUpload rejected', async () => {
        promoCampaignService.uploadPromoCodes.mockImplementation(() => Promise.reject(new Error ('test error')));
        message.error = jest.fn();
        await container.find('UploadPromoCodesModal').prop('onSave')();
        expect(message.error).toHaveBeenCalledWith('test error');
    });
});
