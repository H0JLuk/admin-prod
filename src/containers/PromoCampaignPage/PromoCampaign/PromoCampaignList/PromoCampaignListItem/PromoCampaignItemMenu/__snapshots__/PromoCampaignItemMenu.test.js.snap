// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PromoCampaignItemMenu /> test should match snapshot 1`] = `"<div class=\\"ant-dropdown-trigger menuButton dropdown\\"><span role=\\"img\\" aria-label=\\"ellipsis\\" class=\\"anticon anticon-ellipsis\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" class=\\"\\" data-icon=\\"ellipsis\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M176 511a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0zm280 0a56 56 0 10112 0 56 56 0 10-112 0z\\"></path></svg></span></div>"`;
